core/lib/types/src/contract_verification/contract_identifier.rs

```diff
@@ -58,28 +64,147 @@ impl DetectedMetadata {
     pub fn length(self) -> usize {
         match self {
             DetectedMetadata::Keccak256 => 32,
-            DetectedMetadata::Cbor { full_length } => full_length,
+            DetectedMetadata::Cbor {
+                full_length,
+                metadata: _,
+            } => full_length,
+        }
+    }
+}
+
+/// Represents the compiler version in the Cbor metadata.
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
+#[serde(untagged)]
+pub enum CborCompilerVersion {
+    /// For native solidity and vyper compilers, it is a 3 byte encoding of the compiler version: one byte each for major,
+    /// minor and patch version number. For example, 0.8.28 is encoded as [0, 8, 28].
+    /// More details can be found here:
+    /// https://docs.soliditylang.org/en/latest/metadata.html#encoding-of-the-metadata-hash-in-the-bytecode
+    Native(Vec<u8>),
+    /// For ZKsync solidity compiler, the value consists of semicolon-separated pairs of colon-separated
+    /// compiler names and versions. For example: "zksolc:<version>" or "zksolc:<version>;solc:<version>;llvm:<version>".
+    /// More details can be found here:
+    /// https://matter-labs.github.io/era-compiler-solidity/latest/02-command-line-interface.html#--metadata-hash
+    ///
+    /// For ZKsync vyper compiler, it's "zkvyper:<version>" or "zkvyper:<version>;vyper:<version>".
+    /// More details can be found here:
+    /// https://matter-labs.github.io/era-compiler-vyper/latest/02-command-line-interface.html#--metadata-hash
+    ZKsync(String),
+}
+
+impl CborCompilerVersion {
+    /// Returns the compiler versions from the metadata in a tuple (compiler_version, zk_compiler_version).
+    pub fn get_compiler_versions(&self) -> (Option<String>, Option<String>) {
+        match self {
+            CborCompilerVersion::Native(compiler_version) => {
+                // For native Solc and Vyper compilers, CBOR is a 3 byte encoding of the compiler version: one byte each
+                // for major, minor and patch version number. For example, 0.8.28 is encoded as [0, 8, 28].
```

Comment:
It doesn't have to be 3-byte, [from the docs](https://docs.soliditylang.org/en/latest/metadata.html#encoding-of-the-metadata-hash-in-the-bytecode):

> pre-release builds will instead use a complete version string including commit hash and build date

Reply:
yeah, but we [don't support](https://github.com/matter-labs/zksync-era/blob/main/core/lib/contract_verifier/src/resolver/github/gh_api.rs#L216) pre-releases, do we?
```
            if release.prerelease {
                continue;
            }
```

Reply:
I compiled a contact with a prerelease version and tested, here is what happens in such case:
1. The CBOR data is properly deserialized.
2. When `get_compiler_versions` is called on `CborCompilerVersion`, it returns None for a prerelease version. In such cases, the implementation will fall back to the version from the verification request when compiling the contract.

I added unit tests to cover this use case and added comments to the `get_compiler_versions` fn.